### PR TITLE
[utils] restart download if server does not support byte ranges

### PR DESCRIPTION
**Pull request by remitamine**
to handle the respense of servers which do not support byte ranges.
this happen when using viewster https://github.com/rg3/youtube-dl/issues/6057#issuecomment-126129799.

### CODE REVIEW

File: youtube_dl/utils.py [hunk 1]
```diff
@@ -695,6 +695,8 @@ def http_response(self, req, resp):
             gz = io.BytesIO(self.deflate(resp.read()))
             resp = self.addinfourl_wrapper(gz, old_resp.headers, old_resp.url, old_resp.code)
             resp.msg = old_resp.msg
+        if req.has_header('Range') and resp.getheader('Content-Range') is None:
```
Comment (dstftw) [inline youtube_dl/utils.py]: There is no `getheader` in python 2.

**Comment by dstftw** [comment]
It's probably reasonable to make it even more strict, I mean checking that the lower bound of `Content-Range` header matches the corresponding lower bound of requested `Range` header.

**Comment by remitamine** [comment]
i read the relevant parts in the standard and write this code for checking the length:
```python
def compare_range(req_range, content_range):
    req_range_matchs = re.match('bytes=(\d+)-(\d+)?')
    content_range_matchs = re.match('bytes (\d+)-(\d+)/(\d+|\*)')
    if req_range_matchs is not None and content_range_matchs is not None and req_range_matchs.group(1) == content_range_matchs.group(1):
        if req_range_matchs.group(2) is None and content_range_matchs.group(2) == content_range_matchs.group(3):
            return True
        if req_range_matchs.group(2) is not None and req_range_matchs.group(2) == content_range_matchs.group(2):
            return True
    return False
```
but there is a case in the standard that i don't have urls to check with it's about the cases when the size of the file is not know.

**Comment by dstftw** [comment]
I've only mentioned **lower bound**.

**Comment by remitamine** [comment]
so is this enough:
```python
def compare_range(req_range, content_range):
    req_range_matchs = re.match('bytes=(\d+)-(?:\d+)?')
    content_range_matchs = re.match('bytes (\d+)-\d+/(?:\d+|\*)')
    return req_range_matchs is not None and content_range_matchs is not None and req_range_matchs.group(1) == content_range_matchs.group(1)
```

**Comment by dstftw** [comment]
Basically, you can drop remaining part after dash from regexes, since we are not so intrested in upper bound at all. I would also prefer int comparison.

**Comment by remitamine** [comment]
i fixed some mistakes i made and test it now and resuming files is still working.

**Comment by dstftw** [comment]
I guess there is a better solution. Instead of throwing an exception we could just move the check for `Content-Range` [here](https://github.com/rg3/youtube-dl/blob/e422d7f4f78994de8483d2207ab4e00174a2408c/youtube_dl/downloader/http.py#L60) and just wipe the file if it's invalid. The advantage is that we reuse the same connection and does not throw internal exceptions.
```python
data = self.ydl.urlopen(request)

if resume_len > 0:
    content_range = data.headers.get('Content-Range')
    if content_range:
        content_range_m = re.search(r'bytes (\d+)-', content_range)
        if content_range_m:
            if resume_len != int(content_range_m.group(1)):
                break

resume_len = 0
open_mode = 'wb'
```

**Comment by remitamine** [comment]
there is a problem when i try the code you send always i get error from the server(it work without the change):
```
youtube-dl -v --ignore-config https://www.youtube.com/watch?v=fantexR-qLk
[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['-v', '--ignore-config', 'https://www.youtube.com/watch?v=fantexR-qLk']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2015.07.21
[debug] Python version 3.4.3 - Linux-4.1.2-2-ARCH-x86_64-with-arch-Arch-Linux
[debug] exe versions: ffmpeg 2.7.2, ffprobe 2.7.2, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] fantexR-qLk: Downloading webpage
[youtube] fantexR-qLk: Downloading video info webpage
[youtube] fantexR-qLk: Extracting video information
[youtube] fantexR-qLk: Downloading DASH manifest
[youtube] fantexR-qLk: Downloading DASH manifest
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on 'https://r7---sn-5abxgpxuxaxjvh-55xe.googlevideo.com/videoplayback?id=7da9ed7b147ea8b9&itag=247&source=youtube&requiressl=yes&ms=au&mv=m&pl=17&mn=sn-5abxgpxuxaxjvh-55xe&mm=31&ratebypass=yes&mime=video/webm&gir=yes&clen=10933428&lmt=1438186066563573&dur=92.000&fexp=9406616,9408499,9408710,9415365,9415387,9415485,9415956,9416126,9417269,9417925&sver=3&mt=1438480033&signature=036D05498F55A76BE8994101BBE090EA92EE7BCD.3CC10D80F5267045A56E0DB7181D88488629BE60&upn=DPJhFl8x6_w&key=dg_yt0&ip=197.205.29.203&ipbits=0&expire=1438501714&sparams=ip,ipbits,expire,id,itag,source,requiressl,ms,mv,pl,mn,mm,ratebypass,mime,gir,clen,lmt,dur'
[download] Got server HTTP error. Retrying (attempt 1 of 10)...
[download] Got server HTTP error. Retrying (attempt 2 of 10)...
[download] Got server HTTP error. Retrying (attempt 3 of 10)...
[download] Got server HTTP error. Retrying (attempt 4 of 10)...
[download] Got server HTTP error. Retrying (attempt 5 of 10)...
[download] Got server HTTP error. Retrying (attempt 6 of 10)...
[download] Got server HTTP error. Retrying (attempt 7 of 10)...
[download] Got server HTTP error. Retrying (attempt 8 of 10)...
[download] Got server HTTP error. Retrying (attempt 9 of 10)...
[download] Got server HTTP error. Retrying (attempt 10 of 10)...
ERROR: giving up after 10 retries
  File "/usr/bin/youtube-dl", line 9, in <module>
    load_entry_point('youtube-dl==2015.7.21', 'console_scripts', 'youtube-dl')()
  File "/usr/lib/python3.4/site-packages/youtube_dl/__init__.py", line 410, in main
    _real_main(argv)
  File "/usr/lib/python3.4/site-packages/youtube_dl/__init__.py", line 400, in _real_main
    retcode = ydl.download(all_urls)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 1503, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 667, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 713, in process_ie_result
    return self.process_video_result(ie_result, download=download)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 1173, in process_video_result
    self.process_info(new_info)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 1429, in process_info
    partial_success = dl(fname, new_info)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 1377, in dl
    return fd.download(name, info)
  File "/usr/lib/python3.4/site-packages/youtube_dl/downloader/common.py", line 342, in download
    return self.real_download(filename, info_dict)
  File "/usr/lib/python3.4/site-packages/youtube_dl/downloader/http.py", line 121, in real_download
    self.report_error('giving up after %s retries' % retries)
  File "/usr/lib/python3.4/site-packages/youtube_dl/downloader/common.py", line 155, in report_error
    self.ydl.report_error(*args, **kargs)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 531, in report_error
    self.trouble(error_message, tb)
  File "/usr/lib/python3.4/site-packages/youtube_dl/YoutubeDL.py", line 493, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())
```
i think this shoud be with ==:
```python
if resume_len != int(content_range_m.group(1)):
```
i think if the code is fixed i will close this pull request and you open a pull request with the code you posted here

**Comment by dstftw** [comment]
Note: I've made a correction to the code in couple of minutes after posted.

**Comment by remitamine** [comment]
this code is a lot better than what i use here.
i will close this pull request and you open a pull request with the code you posted here.

**Comment by dstftw** [comment]
Feel free to open it yourself, I'll be unable to test and do it myself until tomorrow.

**Comment by remitamine** [comment]
you get a better idea. you should get credit for it.

**Comment by dstftw** [comment]
Not a big deal, after all we are here to improve the product not to get
some imaginary credits. The plus of PR now would be finished travis build
to analyze the possible consequences till time when I or other developers
will get back to this.

2015-08-02 8:12 GMT+06:00 remitamine <notifications@github.com>:

**Comment by remitamine** [comment]
ok, i will reset the previous commit and put only the last code

**Comment by remitamine** [comment]
because i was working before in the `utils.py` i forget to change it now as the now changes is made to `http.py`.

**Comment by dstftw** [comment]
Cherry-picked in d7d2a9a3dbf1cef78c5085a4aab5d2f336c64cff. Thanks.

**Pull request closed without merging**